bin/fetch-cards-pdf/main.go

```diff
@@ -6,135 +6,221 @@ import (
+		jobQueue <- DownloadJob{
+			RefID: s.AttrOr(":card", ""),
+			Index: 0,
+		}
+		return true
+	})
+}
 
-				err = mw.SetOption("quality", "100")
-				if err != nil {
-					log.Error("setting wand options", "err", err, "option", "quality", "value", "100")
-					return
-				}
+func queueSpecialCaseCards(db *sqlx.DB, jobQueue chan DownloadJob) {
+	log.Debug("looking for special cases")
 
-				err = mw.ReadImageBlob(raw)
-				if err != nil {
-					log.Error("parsing card", "err", err)
-					continue
-				}
+	stmt := `SELECT id, ppid, special FROM refs WHERE special != ""`
+	refs := []reference.Reference{}
+	err := db.Select(&refs, stmt)
+	if err != nil {
+		log.Error("fetch special cases", "err", err)
+		os.Exit(1)
+	}
 
-				err = mw.SetImageFormat("png")
-				if err != nil {
-					log.Error("setting wand format", "err", err)
-					return
-				}
+	for _, ref := range refs {
+		jobQueue <- DownloadJob{
+			RefID: strconv.Itoa(ref.PPID),
+			Index: 1,
+		}
+	}
+}
 
-				mw.SetIteratorIndex(0)
-				// The coordinates here depends on the DPI used?
-				err = mw.CropImage(748, 1050, 140, 227)
-				if err != nil {
-					log.Error("extracting card", "err", err)
-					continue
-				}
+func queueAttachmentCards(db *sqlx.DB, jobQueue chan DownloadJob) {
+	log.Debug("looking for attachements ")
 
-				err = mw.WriteImage(fmt.Sprintf("%s/%s.png", destDir, ref.ID))
-				if err != nil {
-					log.Error("writing card", "err", err)
-					continue
-				}
+	stmt := `
+		SELECT parent.ppid, parent.category_id, parent.special, ref.id
+		FROM (select id, title, linked_to FROM refs WHERE linked_to is not null AND linked_to != 0) as ref
+		LEFT JOIN (select id, ppid, category_id, special FROM refs) AS parent ON ref.linked_to = parent.id
+		`
+	refs := []reference.Reference{}
+	err := db.Select(&refs, stmt)
+	if err != nil {
+		log.Error("fetch attachementss", "err", err)
+		os.Exit(1)
+	}
+
+	for _, ref := range refs {
+		index := 1
+		if ref.CategoryID == 1 || ref.CategoryID == 2 || ref.CategoryID == 10 {
+			// if the ref is a warcaster/warlock/master there is one more card before the attachement
+			index++
+		}
+		if ref.Special != "" {
+			// if the ref if a special case there is one more card before the attachement
+			index++
+		}
+		for _, ref := range refs {
+			jobQueue <- DownloadJob{
+				RefID: strconv.Itoa(ref.PPID),
+				Index: index,
 			}
-		}()
+		}
 	}
+}
 
-	wg.Wait()
+func downloadCard(ppURL, destDir, id string, index int) error {
+	mw := imagick.NewMagickWand()
+	defer mw.Destroy()
+
+	log := log.New("ref", id)
+	log.Debug("retrieving card")
+
+	res, err := http.Get(fmt.Sprintf("%s/?card_items_to_pdf=$%s,1", ppURL, id))
+	if err != nil {
+		return errors.Wrap(err, "retrieving card")
+	}
+
+	raw, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return errors.Wrap(err, "reading card")
+	}
+
+	err = mw.SetOption("density", "300")
+	if err != nil {
+		log.Error("setting wand options", "err", err, "option", "density", "value", "300")
```

Comment:
Either log the error or return it, not both as it forces the caller to either know the details of the function or handle the error twice, which is wasteful in both cases.